texar/torch/data/data/data_iterators.py

```diff
@@ -421,10 +433,16 @@ def __next__(self):
             batch = super().__next__()
             # Drop smaller final batch according to settings. Note that
             # `_batch_size` could be None if dynamic batching is used.
-            if (self._batch_size is not None and
-                    batch.batch_size < self._batch_size and
-                    not self.dataset.hparams.allow_smaller_final_batch):
-                raise StopIteration
+            if _torch_version >= pkg_resources.parse_version("1.3.0"):
```

Comment:
This introduces a large amount of code repetition, plus calling `pkg_resources.parse_version` and checking the version every time is wasteful. I think we can do better. For instance, create a wrapper class over the iterators for PyTorch 1.3 that adds the attribute names with underscore as properties.

It might also be a good idea to move certain classes to a separate file, since this file is getting a bit too long. We might also want to restructure the `data` folder.

Reply:
What I mentioned above is just one way of doing things, and I'm sure there are better ways. Please try to think of a clean and elegant way to resolve this problem.